compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2837,26 +2846,172 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         );
+                        let outer_ty = prev_ty;
+                        let mut assoc_seen = 0;
+                        while let hir::ExprKind::MethodCall(_path_segment, rcvr_expr, _args, span) =
+                            expr.kind
+                        {
+                            // Point at every method call in the chain with the resulting type.
+                            // vec![1, 2, 3].iter().map(mapper).sum<i32>()
+                            //               ^^^^^^ ^^^^^^^^^^^
+                            expr = rcvr_expr;
+
+                            let ocx = ObligationCtxt::new_in_snapshot(self.infcx);
+                            for diff in &type_diffs {
+                                let Sorts(expected_found) = diff else { continue; };
+                                let ty::Projection(proj) = expected_found.expected.kind() else { continue; };
+                                assoc_seen += 1;
+
+                                let origin = TypeVariableOrigin {
+                                    kind: TypeVariableOriginKind::TypeInference,
+                                    span,
+                                };
+                                let trait_def_id = proj.trait_def_id(self.tcx);
+                                let substs =
+                                    InternalSubsts::for_item(self.tcx, trait_def_id, |param, _| {
+                                        match param.kind {
+                                            ty::GenericParamDefKind::Type { .. } => {
+                                                if param.index == 0 {
+                                                    return prev_ty.into();
+                                                }
+                                            }
+                                            ty::GenericParamDefKind::Lifetime
+                                            | ty::GenericParamDefKind::Const { .. } => {}
+                                        }
+                                        self.var_for_def(span, param)
+                                    });
+                                let ty_var = self.infcx.next_ty_var(origin);
+                                let trait_ref = ty::Binder::dummy(ty::PredicateKind::Clause(
+                                    ty::Clause::Projection(ty::ProjectionPredicate {
+                                        projection_ty: ty::ProjectionTy { substs, ..*proj },
+                                        term: ty_var.into(),
+                                    }),
+                                ));
+                                ocx.register_obligation(Obligation::misc(
+                                    self.tcx,
+                                    span,
+                                    expr.hir_id,
+                                    param_env,
+                                    trait_ref,
+                                ));
```

Comment:
You could use `fully_normalize` here. Or just `ocx.normalize` on the projection.

Reply:
I'm not so sure I can use `fully_normalize` because the obligation *must* be included, but I could use `ocx.normalize` though.

Reply:
I'm not sure if I understand -- you could forgo `ocx` and avoid creating an inference variable if you pass the projection (turned into a ty) into `fully_normalize`. Unless I'm misunderstanding what you're doing here, you're just trying to normalize the projection into a concrete value, right?

Reply:
my bad, your snippet made sense